### PR TITLE
LSTMStateTransitionModel (Closes #330) 

### LINKED ISSUE #330: LSTM State Transition Model
# Requirement:
Create tool based on the LSTM algorithm that will use data from one or more run to model state transition (input, state) -> state -> output. 

# D.o.D
- [x] LSTMStateTransitionModel Created
- [x] New dedicated example demonstrating use with ThrownObject generated data and add example to docs
- [x] Tests
- [x] Add dedicated page to docs
- [x] Add to release notes

_Originally posted by @teubert in https://github.com/nasa/prog_models/issues/9#issuecomment-1190780378_

### PR DESCRIPTION
**Pull request by teubert**
First version of the LSTM State Transition Model (Closes #330). 

This is a first iteration of a State Transition Model (i.e., no events) using the LSTM RNN algorithm. Note: this is initial implementation, so there is follow-up work to be done (see bottom). The initial version will take in data from one or more runs to generate a prognostics model which can then be interchanged with a regular prognostics model (full compatibility with prog_algs will be tested later, see https://github.com/nasa/prog_algs/issues/205). 

Our LSTM approach maps from past and current input and past outputs to current output. i.e., 
[z_t-n, u_t-n+1, z_t-n+1, ... z_t-1, u_t] -> z_t
where z is output, u in input, t is current time, n is sequence length (i.e., window size). This is applied repeatedly to predict forward. Past outputs and inputs are stored in the system state. 

This is not very robust - it can only work with simple models. I test this using an example where I build a model with data generated from the ThrownObject model with a set timestep and a different time step. At inspection it matches the behavior well. 

It was unable to match the behavior of more complex data from BatteryElectroChem or BatteryCircuit. I'm hoping some of the work planned in follow-up issues will improve performance enough to capture that. 

### Changes made:
* First draft model
* Example demonstrating capabilities with ThrownObject (#337 for more advanced example in future)
* Simple Tests
* Add dedicated page to sphinx docs
* Add TensorFlow to requirements

### Additional questions for reviewers:
* Does the interface make sense? 
  * We use .from_data() instead of passing data into the constructor. Does that make sense? Maybe I should just use the constructor for this? Thoughts?
  * Should we create a sub package (e.g., data_models) for these?
* Any feedback on approach?
* Ideas for improvement

### Follow-up issues
* #331 
* #332 
* #333 
* #334 
* #335 
* #336 
* #337 
* #338
* #339
* #341 
* https://github.com/nasa/prog_algs/issues/205

### CODE REVIEW

**Comment by teubert** [comment]
Additional questions for reviewers:
* Does the interface make sense? 
  * We use .from_data() instead of passing data into the constructor. Does that make sense? Maybe I should just use the constructor for this? Thoughts?
* Any feedback on approach?
* Ideas for improvement

**Comment by teubert** [comment]
> This pull request **introduces 1 alert** when merging [23511e7](https://github.com/nasa/prog_models/commit/23511e7288cd0804fec153560e6da368ff044c12) into [5bd6a20](https://github.com/nasa/prog_models/commit/5bd6a2074b6444b90a49c024d0745f0a742b261f) - [view on LGTM.com](https://lgtm.com/projects/g/nasa/prog_models/rev/pr-d75a4f5183bc1e54233249d6d634bf50b042f2ed)
> 
> **new alerts:**
> 
> * 1 for `__eq__` not overridden when adding attributes

Fixed in the latest push

File: examples/lstm_model.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    data = m.simulate_to_threshold(future_loading, threshold_keys='impact', save_freq=TIMESTEP, dt=TIMESTEP)
+
+    # Step 2: Generate model
+    # We'll use the LSTMStateTransitionModel class to generate a model from the data.
+    print('Building model...')
+    m2 = LSTMStateTransitionModel.from_data(
+        (data.inputs, data.outputs),  
+        sequence_length=4, 
+        epochs=250,
```
Comment (hlembo) [inline examples/lstm_model.py:43]: What does epochs gets 250 here mean? I assuming that this is part of the LSTMStateTransitionModel?

File: examples/lstm_model.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def future_loading2(t, x = None):
+        # Future Loading is a bit complicated here 
+        # Loading for the resulting model includes the data inputs, 
+        # and the output from the last timestep
+        nonlocal t_counter, x_counter
+        z = m.output(x_counter)
+        z = m2.InputContainer(z.matrix)
+        x_counter = m.next_state(x_counter, future_loading(t), t - t_counter)
+        t_counter = t
+        return z
```
Comment (hlembo) [inline examples/lstm_model.py:60]: Bit complicated for me to personally understand? If you can expalin?

File: setup.py [hunk 3]
```diff
@@ -41,7 +41,8 @@
     install_requires = [
         'scipy',
         'pandas',
-        'matplotlib'
+        'matplotlib',
+        'tensorflow'
     ],
     license = 'NOSA',
```
Comment (hlembo) [inline setup.py:47]: I assume you meant NASA not NOSA here?

File: examples/lstm_model.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        return m.InputContainer({})  # No input for thrown object 
+
+    data = m.simulate_to_threshold(future_loading, threshold_keys='impact', save_freq=TIMESTEP, dt=TIMESTEP)
+
+    # Step 2: Generate model
+    # We'll use the LSTMStateTransitionModel class to generate a model from the data.
+    print('Building model...')
+    m2 = LSTMStateTransitionModel.from_data(
+        (data.inputs, data.outputs),  
+        sequence_length=4,
```
Comment (hlembo) [inline examples/lstm_model.py:42]: Also what is sequence_length used for with LSTM?

File: examples/lstm_model.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    x_counter = m.initialize()
+    def future_loading3(t, x = None):
+        nonlocal t_counter, x_counter
+        z = m3.InputContainer({'x_t-1': x_counter['x'], 'dt': t - t_counter})
+        x_counter = m.next_state(x_counter, future_loading(t), t - t_counter)
+        t_counter = t
+        return z
+
+    # Use new dt, not used in training
+    data = m.simulate_to(data.times[-1], future_loading, dt=TIMESTEP*3, save_freq=TIMESTEP*3)
```
Comment (hlembo) [inline examples/lstm_model.py:121]: Why is this not used in training?

File: setup.py (hunk 3 above)
Comment (teubert) [reply setup.py:47]: Actually it is supposed to be NOSA. NOSA stands for NASA Open Source Agreement

File: examples/lstm_model.py (hunk 5 above)
Comment (teubert) [reply examples/lstm_model.py:123]: added a comments. Does the comment above answer your question?

Comment (hlembo) [reply examples/lstm_model.py:125]: Yep makes more sense now!

File: src/prog_models/lstm_model.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        if params['dropout'] > 0:
+            # Dropout prevents overfitting
+            x = layers.Dropout(params['dropout'])(x)
+
+        x = layers.Dense(z_all.shape[1] if z_all.ndim == 2 else 1)(x)
+        model = keras.Model(inputs, x)
+        model.compile(optimizer="rmsprop", loss="mse", metrics=["mae"])
+        
+        # Train model
+        model.fit(u_all, z_all, epochs=params['epochs'], callbacks = callbacks, validation_split = params['validation_split'])
```
Comment (hlembo) [inline src/prog_models/lstm_model.py:340]: I assume based of the last meeting that this fitness fucntion will be changed. It had something to do with seperate the parts of LSTM so as to not impact and cause the jumpiness we all saw in the output?

**Review by hlembo — APPROVED** [review]

**Review by matteocorbetta — APPROVED** [review]

**Pull request merged**